## **Step 1: Installing Required Libraries**

In [6]:
# ! pip install pandas
# ! pip install numpy 
# ! pip install seaborn
# ! pip install matplotlib.pyplot 
# ! pip install datetime
# ! pip install holidays

## **Step 2: Importing libraries**

In [7]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from datetime import datetime
import holidays


## **Step 3: Reading Datasets**

In [8]:
# Looking at International Departure Report

df1 = pd.read_csv(r"..\..\..\02 Data\Original Data\Internal\International_Report_Departures.csv")
df1

data_dte  Year  Month  usg_apt_id usg_apt  usg_wac  fg_apt_id  \
0       05/01/2006  2006      5       12016     GUM        5      13162   
1       05/01/2003  2003      5       10299     ANC        1      13856   
2       03/01/2007  2007      3       10721     BOS       13      12651   
3       12/01/2004  2004     12       11259     DAL       74      16271   
4       05/01/2009  2009      5       13303     MIA       33      11075   
...            ...   ...    ...         ...     ...      ...        ...   
930803  07/01/2000  2000      7       13303     MIA       33      13605   
930804  04/01/2019  2019      4       13303     MIA       33      14286   
930805  08/01/2000  2000      8       13303     MIA       33      13605   
930806  09/01/2004  2004      9       12266     IAH       74      15632   
930807  05/01/1996  1996      5       12478     JFK       22      14210   

       fg_apt  fg_wac  airlineid carrier  carriergroup        type  Scheduled  \
0         MAJ     844      20177     PFQ             1  Departures          0   
1         OKO     736      20007      5Y             1  Departures          0   
2         KEF     439      20402      GL             1  Departures          0   
3         YYZ     936      20201     AMQ             1  Departures          0   
4         CMW     219      21323      5L             0  Departures          0   
...       ...     ...        ...     ...           ...         ...        ...   
930803    NAS     204      20398      MQ             1  Departures       1422   
930804    PTY     162      19770      CM             0  Departures       1439   
930805    NAS     204      20398      MQ             1  Departures       1441   
930806    VSA     148      19534      AM             0  Departures       1461   
930807    POS     280      19542      BW             0  Departures       2019   

        Charter  Total  
0            10     10  
1            15     15  
2             1      1  
3             1      1  
4            20     20  
...         ...    ...  
930803        0   1422  
930804        0   1439  
930805        0   1441  
930806        0   1461  
930807        0   2019  

[930808 rows x 16 columns]

In [9]:
# Looking at International Passenger Report

df2 = pd.read_csv(r"..\..\..\02 Data\Original Data\Internal\International_Report_Passengers.csv")
df2

data_dte  Year  Month  usg_apt_id usg_apt  usg_wac  fg_apt_id  \
0       05/01/2014  2014      5       14492     RDU       36      11032   
1       06/01/2007  2007      6       13204     MCO       33      16085   
2       12/01/2005  2005     12       11433     DTW       43      10411   
3       04/01/2003  2003      4       13487     MSP       63      16304   
4       12/01/2005  2005     12       12016     GUM        5      11138   
...            ...   ...    ...         ...     ...      ...        ...   
680980  10/01/2015  2015     10       12478     JFK       22      12972   
680981  09/01/2014  2014      9       12478     JFK       22      12972   
680982  10/01/2014  2014     10       12478     JFK       22      12972   
680983  08/01/2014  2014      8       12478     JFK       22      12972   
680984  03/01/1990  1990      3       12478     JFK       22      12972   

       fg_apt  fg_wac  airlineid carrier  carriergroup        type  Scheduled  \
0         CUN     148      19534      AM             0  Passengers          0   
1         YHZ     951      20364      C6             0  Passengers          0   
2         AUA     277      20344      RD             1  Passengers          0   
3         ZIH     148      20204      MG             1  Passengers          0   
4         CRK     766      20312      TZ             1  Passengers          0   
...       ...     ...        ...     ...           ...         ...        ...   
680980    LHR     493      19540      BA             0  Passengers     130217   
680981    LHR     493      19540      BA             0  Passengers     132052   
680982    LHR     493      19540      BA             0  Passengers     132822   
680983    LHR     493      19540      BA             0  Passengers     134263   
680984    LHR     493      19540      BA             0  Passengers     149898   

        Charter   Total  
0           315     315  
1           683     683  
2          1010    1010  
3           508     508  
4            76      76  
...         ...     ...  
680980        0  130217  
680981        0  132052  
680982        0  132822  
680983        0  134263  
680984      297  150195  

[680985 rows x 16 columns]

## **Step 4: Exploring the departure report dataset**

Lets look at the data set and understand the general structure, size and attributes.

In [10]:
# Let's look at first 5 rows of data 

df1.head()

data_dte  Year  Month  usg_apt_id usg_apt  usg_wac  fg_apt_id fg_apt  \
0  05/01/2006  2006      5       12016     GUM        5      13162    MAJ   
1  05/01/2003  2003      5       10299     ANC        1      13856    OKO   
2  03/01/2007  2007      3       10721     BOS       13      12651    KEF   
3  12/01/2004  2004     12       11259     DAL       74      16271    YYZ   
4  05/01/2009  2009      5       13303     MIA       33      11075    CMW   

   fg_wac  airlineid carrier  carriergroup        type  Scheduled  Charter  \
0     844      20177     PFQ             1  Departures          0       10   
1     736      20007      5Y             1  Departures          0       15   
2     439      20402      GL             1  Departures          0        1   
3     936      20201     AMQ             1  Departures          0        1   
4     219      21323      5L             0  Departures          0       20   

   Total  
0     10  
1     15  
2      1  
3      1  
4     20

In [11]:
# Let's look at last 5 rows of data

df1.tail()

data_dte  Year  Month  usg_apt_id usg_apt  usg_wac  fg_apt_id  \
930803  07/01/2000  2000      7       13303     MIA       33      13605   
930804  04/01/2019  2019      4       13303     MIA       33      14286   
930805  08/01/2000  2000      8       13303     MIA       33      13605   
930806  09/01/2004  2004      9       12266     IAH       74      15632   
930807  05/01/1996  1996      5       12478     JFK       22      14210   

       fg_apt  fg_wac  airlineid carrier  carriergroup        type  Scheduled  \
930803    NAS     204      20398      MQ             1  Departures       1422   
930804    PTY     162      19770      CM             0  Departures       1439   
930805    NAS     204      20398      MQ             1  Departures       1441   
930806    VSA     148      19534      AM             0  Departures       1461   
930807    POS     280      19542      BW             0  Departures       2019   

        Charter  Total  
930803        0   1422  
930804        0   1439  
930805        0   1441  
930806        0   1461  
930807        0   2019

In [12]:
# Let's look at the shape of the data

df1.shape

(930808, 16)

In [13]:
# Let's look at the columns in the data

df1.columns

Index(['data_dte', 'Year', 'Month', 'usg_apt_id', 'usg_apt', 'usg_wac',
       'fg_apt_id', 'fg_apt', 'fg_wac', 'airlineid', 'carrier', 'carriergroup',
       'type', 'Scheduled', 'Charter', 'Total'],
      dtype='object')

In [14]:
# Let's look at the data types of the columns

pd.DataFrame(df1.dtypes, columns = ['data_type'])

data_type
data_dte        object
Year             int64
Month            int64
usg_apt_id       int64
usg_apt         object
usg_wac          int64
fg_apt_id        int64
fg_apt          object
fg_wac           int64
airlineid        int64
carrier         object
carriergroup     int64
type            object
Scheduled        int64
Charter          int64
Total            int64

In [15]:
# Let's get general overview of the data

df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 930808 entries, 0 to 930807
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   data_dte      930808 non-null  object
 1   Year          930808 non-null  int64 
 2   Month         930808 non-null  int64 
 3   usg_apt_id    930808 non-null  int64 
 4   usg_apt       930808 non-null  object
 5   usg_wac       930808 non-null  int64 
 6   fg_apt_id     930808 non-null  int64 
 7   fg_apt        930808 non-null  object
 8   fg_wac        930808 non-null  int64 
 9   airlineid     930808 non-null  int64 
 10  carrier       927753 non-null  object
 11  carriergroup  930808 non-null  int64 
 12  type          930808 non-null  object
 13  Scheduled     930808 non-null  int64 
 14  Charter       930808 non-null  int64 
 15  Total         930808 non-null  int64 
dtypes: int64(11), object(5)
memory usage: 113.6+ MB


We can see that the column "carrier" has lesser values as compared to the other columns and might have some null values in it.

In [16]:
# Let's look at descriptive statistic summary of the dataset

df1.describe()

Year          Month     usg_apt_id        usg_wac  \
count  930808.000000  930808.000000  930808.000000  930808.000000   
mean     2006.021361       6.414783   12809.473781      42.511740   
std         8.558831       3.471070    2716.223845      27.571338   
min      1990.000000       1.000000   10010.000000       1.000000   
25%      1999.000000       3.000000   11618.000000      22.000000   
50%      2007.000000       6.000000   12892.000000      33.000000   
75%      2014.000000       9.000000   13487.000000      72.000000   
max      2020.000000      12.000000   99999.000000      93.000000   

           fg_apt_id         fg_wac      airlineid   carriergroup  \
count  930808.000000  930808.000000  930808.000000  930808.000000   
mean    13484.676238     466.910479   20057.217505       0.599361   
std      1932.601107     288.005971     479.071456       0.490028   
min     10119.000000     106.000000   19386.000000       0.000000   
25%     11868.000000     205.000000   19704.000000       0.000000   
50%     13408.000000     427.000000   19991.000000       1.000000   
75%     15084.000000     736.000000   20312.000000       1.000000   
max     16881.000000     975.000000   22067.000000       1.000000   

           Scheduled        Charter          Total  
count  930808.000000  930808.000000  930808.000000  
mean       40.003181       2.005483      42.008665  
std        60.948973       8.278403      60.340835  
min         0.000000       0.000000       1.000000  
25%         0.000000       0.000000       3.000000  
50%        17.000000       0.000000      20.000000  
75%        60.000000       1.000000      60.000000  
max      2019.000000    1092.000000    2019.000000

We can get some general statistical summary about data being from the year 1990 to 2020, minimum and maximum number of scheduled and total departures, their mean and other statistical values to  get a quick glance. 

In [17]:
# Looking at null values and counting them

df1.isnull().sum()

data_dte           0
Year               0
Month              0
usg_apt_id         0
usg_apt            0
usg_wac            0
fg_apt_id          0
fg_apt             0
fg_wac             0
airlineid          0
carrier         3055
carriergroup       0
type               0
Scheduled          0
Charter            0
Total              0
dtype: int64

We can see that the column "carrier" has 3055 null values.

In [18]:
# Looking at the dataframe where carrier column values are null(NaN)

df1[df1['carrier'].isnull()]

data_dte  Year  Month  usg_apt_id usg_apt  usg_wac  fg_apt_id  \
21      06/01/2005  2005      6       12478     JFK       22      12991   
37      11/01/2006  2006     11       12478     JFK       22      14210   
255     07/01/2009  2009      7       15070     SWF       22      12957   
267     01/01/2006  2006      1       12478     JFK       22      15115   
318     01/01/2006  2006      1       11697     FLL       33      14210   
...            ...   ...    ...         ...     ...      ...        ...   
930487  05/01/1999  1999      5       14843     SJU        3      11519   
930495  03/01/1999  1999      3       14843     SJU        3      11519   
930506  03/01/1996  1996      3       14843     SJU        3      11519   
930608  08/01/1999  1999      8       14843     SJU        3      11519   
930618  07/01/1999  1999      7       14843     SJU        3      11519   

       fg_apt  fg_wac  airlineid carrier  carriergroup        type  Scheduled  \
21        LIS     469      20415     NaN             1  Departures          0   
37        POS     280      20415     NaN             1  Departures          0   
255       LGG     409      20415     NaN             1  Departures          0   
267       TAB     280      20415     NaN             1  Departures          0   
318       POS     280      20415     NaN             1  Departures          0   
...       ...     ...        ...     ...           ...         ...        ...   
930487    EIS     282      20414     NaN             1  Departures        682   
930495    EIS     282      20414     NaN             1  Departures        684   
930506    EIS     282      20414     NaN             1  Departures        686   
930608    EIS     282      20414     NaN             1  Departures        734   
930618    EIS     282      20414     NaN             1  Departures        745   

        Charter  Total  
21            1      1  
37           10     10  
255           1      1  
267           2      2  
318          10     10  
...         ...    ...  
930487        0    682  
930495        0    684  
930506        0    686  
930608        0    734  
930618        0    745  

[3055 rows x 16 columns]

In [19]:
# Seeing if we can find carrier null values from exisiting data

# Finding unique value counts for airline ids where carrier is null

df1[df1["carrier"].isna()]["airlineid"].value_counts().rename_axis('airline_id').reset_index(name='counts')

airline_id  counts
0       20414    1609
1       20415    1446

In [20]:
#  Finding the carrier where airline id is 20414 

df1.loc[df1["airlineid"] == 20414, ["airlineid", "carrier"]].head(1)


airlineid carrier
264      20414      OW

In [21]:
#  Finding the carrier where airline id is 20415
 
df1.loc[df1["airlineid"] == 20415, ["airlineid", "carrier"]].head(5)


airlineid carrier
21       20415     NaN
37       20415     NaN
103      20415      XG
156      20415      XG
255      20415     NaN

### Notes:
* We have 16 columns with 11 integer and 5 object data types.
* We might need to change the names and types of some of the columns, e.g date, possibly carrier and so on for readability and consistency.
* There is one column named "carrier" which has 3055 null values and we are able to find those missing values based on the corresponding airline id which we will replace them further in the cleaning process.

## **Step 5: Exploring the Passengers Report Dataset**

Here, we will repeat all the steps like above to look at the passenger report data set and understand it's general structure, size and attributes. 

In [22]:
# Let's look at first 5 rows of data 

df2.head()

data_dte  Year  Month  usg_apt_id usg_apt  usg_wac  fg_apt_id fg_apt  \
0  05/01/2014  2014      5       14492     RDU       36      11032    CUN   
1  06/01/2007  2007      6       13204     MCO       33      16085    YHZ   
2  12/01/2005  2005     12       11433     DTW       43      10411    AUA   
3  04/01/2003  2003      4       13487     MSP       63      16304    ZIH   
4  12/01/2005  2005     12       12016     GUM        5      11138    CRK   

   fg_wac  airlineid carrier  carriergroup        type  Scheduled  Charter  \
0     148      19534      AM             0  Passengers          0      315   
1     951      20364      C6             0  Passengers          0      683   
2     277      20344      RD             1  Passengers          0     1010   
3     148      20204      MG             1  Passengers          0      508   
4     766      20312      TZ             1  Passengers          0       76   

   Total  
0    315  
1    683  
2   1010  
3    508  
4     76

In [23]:
# Let's look at last 5 rows of data

df2.tail()

data_dte  Year  Month  usg_apt_id usg_apt  usg_wac  fg_apt_id  \
680980  10/01/2015  2015     10       12478     JFK       22      12972   
680981  09/01/2014  2014      9       12478     JFK       22      12972   
680982  10/01/2014  2014     10       12478     JFK       22      12972   
680983  08/01/2014  2014      8       12478     JFK       22      12972   
680984  03/01/1990  1990      3       12478     JFK       22      12972   

       fg_apt  fg_wac  airlineid carrier  carriergroup        type  Scheduled  \
680980    LHR     493      19540      BA             0  Passengers     130217   
680981    LHR     493      19540      BA             0  Passengers     132052   
680982    LHR     493      19540      BA             0  Passengers     132822   
680983    LHR     493      19540      BA             0  Passengers     134263   
680984    LHR     493      19540      BA             0  Passengers     149898   

        Charter   Total  
680980        0  130217  
680981        0  132052  
680982        0  132822  
680983        0  134263  
680984      297  150195

In [24]:
# Let's look at the shape of the data

df2.shape

(680985, 16)

In [25]:
# Let's look at the columns in the data

df2.columns

Index(['data_dte', 'Year', 'Month', 'usg_apt_id', 'usg_apt', 'usg_wac',
       'fg_apt_id', 'fg_apt', 'fg_wac', 'airlineid', 'carrier', 'carriergroup',
       'type', 'Scheduled', 'Charter', 'Total'],
      dtype='object')

We can see that the number of columns and column names are same like the departure report dataset.

In [26]:
# Let's look at the data types of each of the columns

pd.DataFrame(df2.dtypes, columns = ['data_type'])

data_type
data_dte        object
Year             int64
Month            int64
usg_apt_id       int64
usg_apt         object
usg_wac          int64
fg_apt_id        int64
fg_apt          object
fg_wac           int64
airlineid        int64
carrier         object
carriergroup     int64
type            object
Scheduled        int64
Charter          int64
Total            int64

In [27]:
# Let's get general overview of the data

df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 680985 entries, 0 to 680984
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   data_dte      680985 non-null  object
 1   Year          680985 non-null  int64 
 2   Month         680985 non-null  int64 
 3   usg_apt_id    680985 non-null  int64 
 4   usg_apt       680985 non-null  object
 5   usg_wac       680985 non-null  int64 
 6   fg_apt_id     680985 non-null  int64 
 7   fg_apt        680985 non-null  object
 8   fg_wac        680985 non-null  int64 
 9   airlineid     680985 non-null  int64 
 10  carrier       678188 non-null  object
 11  carriergroup  680985 non-null  int64 
 12  type          680985 non-null  object
 13  Scheduled     680985 non-null  int64 
 14  Charter       680985 non-null  int64 
 15  Total         680985 non-null  int64 
dtypes: int64(11), object(5)
memory usage: 83.1+ MB


In [28]:
# Let's look at descriptive statistic summary of the dataset

df2.describe()

Year          Month     usg_apt_id        usg_wac  \
count  680985.000000  680985.000000  680985.000000  680985.000000   
mean     2006.141110       6.366576   12779.103825      44.394527   
std         8.716122       3.466745    2510.939346      28.110753   
min      1990.000000       1.000000   10010.000000       1.000000   
25%      1999.000000       3.000000   11618.000000      22.000000   
50%      2007.000000       6.000000   12889.000000      33.000000   
75%      2014.000000       9.000000   13796.000000      74.000000   
max      2020.000000      12.000000   99999.000000      93.000000   

           fg_apt_id         fg_wac      airlineid   carriergroup  \
count  680985.000000  680985.000000  680985.000000  680985.000000   
mean    13493.331869     468.122264   20031.610461       0.562233   
std      1955.917890     292.799407     510.270976       0.496112   
min     10125.000000     106.000000   19386.000000       0.000000   
25%     11760.000000     204.000000   19623.000000       0.000000   
50%     13408.000000     429.000000   19917.000000       1.000000   
75%     15092.000000     736.000000   20312.000000       1.000000   
max     16881.000000     975.000000   22067.000000       1.000000   

           Scheduled        Charter          Total  
count  680985.000000  680985.000000  680985.000000  
mean     6478.893801     196.740288    6675.634089  
std      8622.704655     954.759814    8532.400269  
min         0.000000       0.000000       1.000000  
25%       142.000000       0.000000     633.000000  
50%      3561.000000       0.000000    3803.000000  
75%      9700.000000       0.000000    9761.000000  
max    149898.000000   58284.000000  150195.000000

In [29]:
# Looking at null values and counting them

df2.isnull().sum()

data_dte           0
Year               0
Month              0
usg_apt_id         0
usg_apt            0
usg_wac            0
fg_apt_id          0
fg_apt             0
fg_wac             0
airlineid          0
carrier         2797
carriergroup       0
type               0
Scheduled          0
Charter            0
Total              0
dtype: int64

In [30]:
# Looking at the dataframe where carrier column values are null(NaN)

df2[df2['carrier'].isnull()]

data_dte  Year  Month  usg_apt_id usg_apt  usg_wac  fg_apt_id  \
50      01/01/2006  2006      1       11618     EWR       21      14293   
83      06/01/2007  2007      6       12478     JFK       22      14210   
276     05/01/2006  2006      5       12478     JFK       22      14210   
391     05/01/2009  2009      5       11298     DFW       74      14210   
397     02/01/2005  2005      2       12478     JFK       22      10575   
...            ...   ...    ...         ...     ...      ...        ...   
640284  06/01/1999  1999      6       14843     SJU        3      14739   
641266  07/01/1999  1999      7       14843     SJU        3      11519   
643130  03/01/1997  1997      3       14843     SJU        3      11519   
644862  03/01/1999  1999      3       14843     SJU        3      11519   
652274  07/01/1999  1999      7       14843     SJU        3      14739   

       fg_apt  fg_wac  airlineid carrier  carriergroup        type  Scheduled  \
50        PUJ     224      20415     NaN             1  Passengers          0   
83        POS     280      20415     NaN             1  Passengers          0   
276       POS     280      20415     NaN             1  Passengers          0   
391       POS     280      20415     NaN             1  Passengers          0   
397       BGI     205      20415     NaN             1  Passengers          0   
...       ...     ...        ...     ...           ...         ...        ...   
640284    SDQ     224      20414     NaN             1  Passengers      20655   
641266    EIS     282      20414     NaN             1  Passengers      20872   
643130    EIS     282      20414     NaN             1  Passengers      21326   
644862    EIS     282      20414     NaN             1  Passengers      21771   
652274    SDQ     224      20414     NaN             1  Passengers      24240   

        Charter  Total  
50          231    231  
83         2159   2159  
276        4624   4624  
391          64     64  
397        1680   1680  
...         ...    ...  
640284        0  20655  
641266        0  20872  
643130        0  21326  
644862        0  21771  
652274        0  24240  

[2797 rows x 16 columns]

In [31]:
# Seeing if we can find carrier null values from the exisiting data

# Finding unique value counts for airline ids where carrier is null

pd.DataFrame(df2[df2["carrier"].isna()]["airlineid"].value_counts())

airlineid
20414       1603
20415       1194

The same airline ids of '20414' and '20415' are having some null values for carriers in this dataset as well. 

In [32]:
#  Finding carriers for airline id 20414 and 20415

print(df2[df2["airlineid"] == 20414]["carrier"].value_counts())
print(df2[df2["airlineid"] == 20415]["carrier"].value_counts())


OW    4567
Name: carrier, dtype: int64
XG    1887
Name: carrier, dtype: int64


### Notes:
* It's a similar dataframe just like the departure report where we have 16 columns with 11 integer and 5 object data types.
* We might need to change the names and types of some of the columns, e.g date, possibly carrier and so on for readability and consistency.
* There are 2797 null values in the "carrier" column.
* we are able to find those missing values based on the corresponding airline id which we will replace them further in the cleaning process.

## **Step 6: Cleaning the departure report dataset**

#### Required Data clean up
* Find and remove duplicate rows
* Replace missing values in the **carrier** column with the corresponding values found above.
* Change data types of **carrier** to **string** and **Data_dte** to **date**
* Convert column names to lower case and rename a few columns for readability.

### *Departure Report Dataset*

#### Looking at duplicate values

In [33]:
# Looking at duplicate rows in the Departures report 

df1[df1.duplicated()]

Empty DataFrame
Columns: [data_dte, Year, Month, usg_apt_id, usg_apt, usg_wac, fg_apt_id, fg_apt, fg_wac, airlineid, carrier, carriergroup, type, Scheduled, Charter, Total]
Index: []

There are no duplicate rows present in this dataset.

#### Converting the column names to lower case

In [34]:
df1.columns = df1.columns.str.lower()
df1.head()

data_dte  year  month  usg_apt_id usg_apt  usg_wac  fg_apt_id fg_apt  \
0  05/01/2006  2006      5       12016     GUM        5      13162    MAJ   
1  05/01/2003  2003      5       10299     ANC        1      13856    OKO   
2  03/01/2007  2007      3       10721     BOS       13      12651    KEF   
3  12/01/2004  2004     12       11259     DAL       74      16271    YYZ   
4  05/01/2009  2009      5       13303     MIA       33      11075    CMW   

   fg_wac  airlineid carrier  carriergroup        type  scheduled  charter  \
0     844      20177     PFQ             1  Departures          0       10   
1     736      20007      5Y             1  Departures          0       15   
2     439      20402      GL             1  Departures          0        1   
3     936      20201     AMQ             1  Departures          0        1   
4     219      21323      5L             0  Departures          0       20   

   total  
0     10  
1     15  
2      1  
3      1  
4     20

### Rename columns

In [35]:
df1.rename(columns={'data_dte':'date','airlineid':'airline_id','carriergroup':'carrier_group'}, inplace = True)
df1.head()

date  year  month  usg_apt_id usg_apt  usg_wac  fg_apt_id fg_apt  \
0  05/01/2006  2006      5       12016     GUM        5      13162    MAJ   
1  05/01/2003  2003      5       10299     ANC        1      13856    OKO   
2  03/01/2007  2007      3       10721     BOS       13      12651    KEF   
3  12/01/2004  2004     12       11259     DAL       74      16271    YYZ   
4  05/01/2009  2009      5       13303     MIA       33      11075    CMW   

   fg_wac  airline_id carrier  carrier_group        type  scheduled  charter  \
0     844       20177     PFQ              1  Departures          0       10   
1     736       20007      5Y              1  Departures          0       15   
2     439       20402      GL              1  Departures          0        1   
3     936       20201     AMQ              1  Departures          0        1   
4     219       21323      5L              0  Departures          0       20   

   total  
0     10  
1     15  
2      1  
3      1  
4     20

### Changing data types

In [36]:
# Changing date from object data type to a date format

df1["date"] = pd.to_datetime(df1["date"])
df1.head()

date  year  month  usg_apt_id usg_apt  usg_wac  fg_apt_id fg_apt  \
0 2006-05-01  2006      5       12016     GUM        5      13162    MAJ   
1 2003-05-01  2003      5       10299     ANC        1      13856    OKO   
2 2007-03-01  2007      3       10721     BOS       13      12651    KEF   
3 2004-12-01  2004     12       11259     DAL       74      16271    YYZ   
4 2009-05-01  2009      5       13303     MIA       33      11075    CMW   

   fg_wac  airline_id carrier  carrier_group        type  scheduled  charter  \
0     844       20177     PFQ              1  Departures          0       10   
1     736       20007      5Y              1  Departures          0       15   
2     439       20402      GL              1  Departures          0        1   
3     936       20201     AMQ              1  Departures          0        1   
4     219       21323      5L              0  Departures          0       20   

   total  
0     10  
1     15  
2      1  
3      1  
4     20

In [37]:
# Changing carrier column from object to string

df1['carrier'] = df1['carrier'].astype(str)

### Replacing null values

#### Departure Report Dataset

In [38]:
# Replacing null values in the carrier column of the dataset with values correseponding to the airline id.

# Filling missing values for 20414 with "OW" - see previous cell

df1.loc[df1["airline_id"] == 20414, ["carrier"]].fillna("OW", inplace=True)

In [39]:
# Filling missing values for 20415 with "XG" - see previous cell

df1.loc[df1["airline_id"] == 20415, ["carrier"]].fillna("XG", inplace=True)

# Another way- df1.loc[(df1["airline_id"] == 20415) & df1["carrier"].isnull(), ["carrier"]] = "XG"


In [40]:
# Rechecking if the null values are replaced

df1.isnull().sum()

date             0
year             0
month            0
usg_apt_id       0
usg_apt          0
usg_wac          0
fg_apt_id        0
fg_apt           0
fg_wac           0
airline_id       0
carrier          0
carrier_group    0
type             0
scheduled        0
charter          0
total            0
dtype: int64

#### Replacing the values of 1 and 0 in the carrier_group with "Domestic"and "Foreign"

In [41]:
# First let's change carrier_group column to string

df1['carrier_group'] = df1['carrier_group'].astype(str) 

In [42]:
# Let's now change values in the column from 1 to Domestic.

df1['carrier_group'] = df1['carrier_group'].replace(str(1),'Domestic')

In [43]:
# Let's change the values in the column from 0 to Foreign

df1['carrier_group'] = df1['carrier_group'].replace(str(0),'Foreign')

In [44]:
# Let's look at the carrier_group column to verify the changed values

df1['carrier_group']

0         Domestic
1         Domestic
2         Domestic
3         Domestic
4          Foreign
            ...   
930803    Domestic
930804     Foreign
930805    Domestic
930806     Foreign
930807     Foreign
Name: carrier_group, Length: 930808, dtype: object

### *Passenger Report Dataset*

Repeating the same above steps for cleaning in the 2nd dataset

#### Looking at duplicate rows in the dataset

In [45]:
df2[df2.duplicated()]

Empty DataFrame
Columns: [data_dte, Year, Month, usg_apt_id, usg_apt, usg_wac, fg_apt_id, fg_apt, fg_wac, airlineid, carrier, carriergroup, type, Scheduled, Charter, Total]
Index: []

No duplicate rows found in the dataset.

#### Changing column names to lower case

In [46]:
df2.columns = df2.columns.str.lower()
df2.head()

data_dte  year  month  usg_apt_id usg_apt  usg_wac  fg_apt_id fg_apt  \
0  05/01/2014  2014      5       14492     RDU       36      11032    CUN   
1  06/01/2007  2007      6       13204     MCO       33      16085    YHZ   
2  12/01/2005  2005     12       11433     DTW       43      10411    AUA   
3  04/01/2003  2003      4       13487     MSP       63      16304    ZIH   
4  12/01/2005  2005     12       12016     GUM        5      11138    CRK   

   fg_wac  airlineid carrier  carriergroup        type  scheduled  charter  \
0     148      19534      AM             0  Passengers          0      315   
1     951      20364      C6             0  Passengers          0      683   
2     277      20344      RD             1  Passengers          0     1010   
3     148      20204      MG             1  Passengers          0      508   
4     766      20312      TZ             1  Passengers          0       76   

   total  
0    315  
1    683  
2   1010  
3    508  
4     76

#### Changing a few column names

In [47]:
df2.rename(columns={'data_dte':'date','airlineid':'airline_id','carriergroup':'carrier_group'},inplace=True)
df2.head()

date  year  month  usg_apt_id usg_apt  usg_wac  fg_apt_id fg_apt  \
0  05/01/2014  2014      5       14492     RDU       36      11032    CUN   
1  06/01/2007  2007      6       13204     MCO       33      16085    YHZ   
2  12/01/2005  2005     12       11433     DTW       43      10411    AUA   
3  04/01/2003  2003      4       13487     MSP       63      16304    ZIH   
4  12/01/2005  2005     12       12016     GUM        5      11138    CRK   

   fg_wac  airline_id carrier  carrier_group        type  scheduled  charter  \
0     148       19534      AM              0  Passengers          0      315   
1     951       20364      C6              0  Passengers          0      683   
2     277       20344      RD              1  Passengers          0     1010   
3     148       20204      MG              1  Passengers          0      508   
4     766       20312      TZ              1  Passengers          0       76   

   total  
0    315  
1    683  
2   1010  
3    508  
4     76

#### Changing data types

In [48]:
# Changing date from object type to datetime format

df2["date"] = pd.to_datetime(df2["date"])
df2.head()

date  year  month  usg_apt_id usg_apt  usg_wac  fg_apt_id fg_apt  \
0 2014-05-01  2014      5       14492     RDU       36      11032    CUN   
1 2007-06-01  2007      6       13204     MCO       33      16085    YHZ   
2 2005-12-01  2005     12       11433     DTW       43      10411    AUA   
3 2003-04-01  2003      4       13487     MSP       63      16304    ZIH   
4 2005-12-01  2005     12       12016     GUM        5      11138    CRK   

   fg_wac  airline_id carrier  carrier_group        type  scheduled  charter  \
0     148       19534      AM              0  Passengers          0      315   
1     951       20364      C6              0  Passengers          0      683   
2     277       20344      RD              1  Passengers          0     1010   
3     148       20204      MG              1  Passengers          0      508   
4     766       20312      TZ              1  Passengers          0       76   

   total  
0    315  
1    683  
2   1010  
3    508  
4     76

In [49]:
# Changing carrier column from object to string

df2["carrier"] = df2["carrier"].astype(str)
df2.head()

date  year  month  usg_apt_id usg_apt  usg_wac  fg_apt_id fg_apt  \
0 2014-05-01  2014      5       14492     RDU       36      11032    CUN   
1 2007-06-01  2007      6       13204     MCO       33      16085    YHZ   
2 2005-12-01  2005     12       11433     DTW       43      10411    AUA   
3 2003-04-01  2003      4       13487     MSP       63      16304    ZIH   
4 2005-12-01  2005     12       12016     GUM        5      11138    CRK   

   fg_wac  airline_id carrier  carrier_group        type  scheduled  charter  \
0     148       19534      AM              0  Passengers          0      315   
1     951       20364      C6              0  Passengers          0      683   
2     277       20344      RD              1  Passengers          0     1010   
3     148       20204      MG              1  Passengers          0      508   
4     766       20312      TZ              1  Passengers          0       76   

   total  
0    315  
1    683  
2   1010  
3    508  
4     76

#### Replacing null values

In [50]:
# Replacing null values in the carrier column of the dataset with values correseponding to the airline id.

# Filling missing values for 20414 with "OW" - see previous cell

df2.loc[df2["airline_id"] == 20414, ["carrier"]].fillna("OW", inplace=True)

In [51]:
# Filling missing values for 20415 with "XG" - see previous cell

df2.loc[df2["airline_id"] == 20415, ["carrier"]].fillna("XG", inplace=True)

In [52]:
# Rechecking null values in the dataset

df2.isnull().sum()

date             0
year             0
month            0
usg_apt_id       0
usg_apt          0
usg_wac          0
fg_apt_id        0
fg_apt           0
fg_wac           0
airline_id       0
carrier          0
carrier_group    0
type             0
scheduled        0
charter          0
total            0
dtype: int64

#### Replacing the values of 1 and 0 in the carrier_group with "Domestic"and "Foreign"

In [53]:
# changing carrier_group column to string

df2['carrier_group'] = df2['carrier_group'].astype(str) 

In [54]:
 # changing values in the column from 1 to Domestic.

df2['carrier_group'] = df2['carrier_group'].replace(str(1),'Domestic')

In [55]:
# changing values in the column from 0 to Foreign.

df2['carrier_group'] = df2['carrier_group'].replace(str(0),'Foreign')

In [56]:
# Looking at the carrier_group column to check if values are replaced

df2['carrier_group']

0          Foreign
1          Foreign
2         Domestic
3         Domestic
4         Domestic
            ...   
680980     Foreign
680981     Foreign
680982     Foreign
680983     Foreign
680984     Foreign
Name: carrier_group, Length: 680985, dtype: object